src/query/storages/common/table-meta/src/meta/v4/segment.rs

```diff
@@ -0,0 +1,285 @@
+    pub format_version: FormatVersion,
+    /// blocks belong to this segment
+    pub blocks: Vec<Arc<BlockMeta>>,
+    /// summary statistics
+    pub summary: Statistics,
+}
+
+impl SegmentInfo {
+    pub fn new(blocks: Vec<Arc<BlockMeta>>, summary: Statistics) -> Self {
+        Self {
+            format_version: SegmentInfo::VERSION,
+            blocks,
+            summary,
+        }
+    }
+
+    // Total block bytes of this segment.
+    pub fn total_bytes(&self) -> u64 {
+        self.blocks.iter().map(|v| v.block_size).sum()
+    }
+
+    #[inline]
+    pub fn encoding() -> MetaEncoding {
+        MetaEncoding::MessagePack
+    }
+
+    // Encode self.blocks as RawBlockMeta.
+    fn block_raw_bytes(&self) -> Result<RawBlockMeta> {
+        let encoding = MetaEncoding::MessagePack;
+        let bytes = encode(&encoding, &self.blocks)?;
+
+        let compression = MetaCompression::default();
+        let compressed = compress(&compression, bytes)?;
+
+        Ok(RawBlockMeta {
+            bytes: compressed,
+            encoding,
+            compression,
+        })
+    }
+}
+
+// use the chain of converters, for versions before v3
+impl<T> From<T> for SegmentInfo
+where T: Into<v2::SegmentInfo>
+{
+    fn from(value: T) -> Self {
+        Self::from_v2(value.into())
+    }
+}
+
+impl From<v3::SegmentInfo> for SegmentInfo {
+    fn from(value: v3::SegmentInfo) -> Self {
+        Self::from_v3(value)
+    }
+}
+
+impl SegmentInfo {
+    pub fn from_v3(s: v3::SegmentInfo) -> Self {
+        // NOTE: it is important to let the format_version return from here
+        // carries the format_version of segment info being converted.
+        Self {
+            format_version: s.format_version,
+            blocks: s.blocks.into_iter().map(|v| Arc::new(v.into())).collect(),
+            summary: s.summary.into(),
+        }
+    }
+    pub fn from_v2(s: v2::SegmentInfo) -> Self {
+        // NOTE: it is important to let the format_version return from here
+        // carries the format_version of segment info being converted.
+        Self {
+            format_version: s.format_version,
+            blocks: s.blocks,
+            summary: s.summary,
+        }
+    }
+
+    /// Serializes the Segment struct to a byte vector.
+    ///
+    /// The byte vector contains the format version, encoding, compression, and compressed block data and
+    /// summary data. The encoding and compression are set to default values. The block data and summary
+    /// data are encoded and compressed, respectively.
+    ///
+    /// # Returns
+    ///
+    /// A Result containing the serialized Segment data as a byte vector. If any errors occur during
+    /// encoding, compression, or writing to the byte vector, an error will be returned.
+    pub fn to_bytes(&self) -> Result<Vec<u8>> {
+        self.to_bytes_with_encoding(MetaEncoding::MessagePack)
+    }
+
+    pub fn to_bytes_with_encoding(&self, encoding: MetaEncoding) -> Result<Vec<u8>> {
```

Comment:
this should not be `pub` in case someone use `to_bytes_with_encoding(MetaEncoding::Bincode)` directly

Reply:
this method is used by the benchmark code. let me comment out the bench_function